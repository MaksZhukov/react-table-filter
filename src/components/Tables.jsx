import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import 'rc-checkbox/assets/index.css';

class Tables extends PureComponent {
  render() {
    const { tables, checkedAllCells } = this.props;
    return (
      <div className="tables">
        {tables.length >= 0 && !checkedAllCells
          && tables.map(table => (
            <table key={table} className="table">
              <caption className="table-caption">
                {table.name}
              </caption>
              <thead>
                <tr>
                  {table.dimensions.map(dimension => (
                    <th key={dimension}>
                      {dimension}
                    </th>
                  ))}
                </tr>
              </thead>
              <tbody>
                {table.rows.map((row, rowKey) => (
                  <tr key={rowKey.toString()}>
                    {row.map((cell, keyCell) => (
                      <td key={keyCell.toString()}>
                        {cell && cell.value}
                      </td>
                    ))}
                  </tr>
                ))}
              </tbody>
            </table>
          ))
        }
      </div>
    );
  }
}

Tables.propTypes = {
  tables: PropTypes.instanceOf(Array).isRequired,
  checkedAllCells: PropTypes.bool.isRequired,
};

export default Tables;
