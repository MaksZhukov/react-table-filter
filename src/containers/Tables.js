import { connect } from 'react-redux';
import Tables from '../components/Tables';

const mapStateToProps = (state, {
  id,
}) => {
  const panel = state.get('panels').find(pan => pan.get('id') === id);
  const cells = panel.getIn(['filters', 'cells']).filter(cell => cell.get('checked')).toJS();
  const contexts = panel.getIn(['filters', 'contexts']).toJS();
  const dimensions = panel.getIn(['filters', 'dimensions']).toJS();
  const checkedAllCells = panel.getIn(['filters', 'checkedAllCells']);
  const tables = [];
  contexts.forEach((table) => {
    if (cells.filter(cell => cell.parents.context === table).length !== 0) {
      const selectedDimensions = [];
      const selectedCells = [];
      dimensions.forEach((dimension) => {
        const cellsInTheDim = cells.filter(c => c.parents.dimension === dimension
          && c.parents.context === table);
        if (cellsInTheDim.length !== 0) {
          selectedDimensions.push(dimension);
          selectedCells.push(cellsInTheDim);
        }
      });
      const maxLength = Math.max(...selectedCells.map(cell => cell.length));
      const rows = Array(maxLength).fill(null).map(() => []);
      rows.forEach((row, keyRow) => {
        selectedCells.forEach((cell, keyCell) => {
          row.push(selectedCells[keyCell][keyRow]);
        });
      });
      tables.push({
        name: table,
        dimensions: selectedDimensions,
        rows,
      });
    }
  });
  return { tables, checkedAllCells };
};

export default connect(mapStateToProps)(Tables);
