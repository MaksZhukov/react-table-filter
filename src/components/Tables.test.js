import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallowToJson } from 'enzyme-to-json';
import Tables from './Tables';


configure({ adapter: new Adapter() });
describe('component Tables', () => {
  const props = {
    tables: [{
      name: 'table1',
      dimensions: ['dim1'],
      rows: [[{ value: 'cell1' }]],
    }],
    checkedAllCells: false,
  };
  it('render component Tables without crashing', () => {
    const wrapper = shallow(<Tables {...props} />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
  it('dont render tables because checkedAllCells equal true', () => {
    const wrapper = shallow(<Tables {...{ ...props, checkedAllCells: true }} />);
    expect(wrapper.find('table').length).toBe(0);
  });
  it('table component has two props', () => {
    const wrapper = shallow(<Tables {...props} />);
    expect(Object.keys(wrapper.instance().props).length).toBe(2);
  });
});
