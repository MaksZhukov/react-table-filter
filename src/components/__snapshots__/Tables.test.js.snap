// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component Tables render component Tables without crashing 1`] = `
<div
  className="tables"
>
  <table
    className="table"
    key="[object Object]"
  >
    <caption
      className="table-caption"
    >
      table1
    </caption>
    <thead>
      <tr>
        <th
          key="dim1"
        >
          dim1
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        key="0"
      >
        <td
          key="0"
        >
          cell1
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
